ex2_hal/iris/hardware_interface/source/iris.c

```diff
@@ -0,0 +1,460 @@
+            case SEND_COMMAND:
+            {
+                ret = send_command(IRIS_GET_IMAGE_LENGTH);
+                if (ret == IRIS_ACK) {
+                    controller_state = GET_DATA;
+                } else {
+                    controller_state = FINISH;
+                }
+                break;
+            }
+            case GET_DATA:
+            {
+                uint16_t * image_length_buffer = (uint16_t*) pvPortMalloc(MAX_IMAGE_LENGTH * sizeof(uint16_t));
+                memset(image_length_buffer, 0, MAX_IMAGE_LENGTH);
+                if (image_length_buffer == NULL) {
+                    ex2_log("Failed attempt to dynamically allocate memory under iris get image length");
+                    return IRIS_HAL_ERROR;;
+                }
+                ret = get_data(image_length_buffer, MAX_IMAGE_LENGTH);
+                if (ret == IRIS_HAL_OK) {
+                    controller_state = FINISH;
+                } else {
+                    controller_state = ERROR_STATE;
+                }
+                /* It is expected that the first byte in the buffer will be the LSB */
+                *(image_length) = (uint32_t)((uint8_t)image_length_buffer[2]<<16 | (uint8_t)image_length_buffer[1]<<8 | (uint8_t)image_length_buffer[0]); // Concatenate image_length_buffer
+
+                free(image_length_buffer);
+                controller_state = FINISH;
+                break;
+            }
+            case FINISH:
+            {
+                ex2_log("Iris returns ACK on get image length command");
+                return IRIS_HAL_OK;
+            }
+            case ERROR_STATE:
+            {
+                ex2_log("Iris returns NACK on get image length command");
+                return IRIS_HAL_ERROR;
+            }
+        }
+    }
+    return IRIS_HAL_ERROR;
+}
+
+/**
+ * @brief
+ *   Sends a transfer image command to Iris, and expects to receive the
+ *   image data of size image_length
+ *
+ * @param[in] image_length
+ *   Number of bytes expected to be received from Iris
+ *
+ * @return
+ *   Returns IRIS_HAL_OK if equipment handler returns IRIS_ACK, else IRIS_HAL_ERROR
+ **/
+IrisHALReturn iris_transfer_image(uint32_t image_length) {
+    uint32_t num_transfer;
+    IrisLowLevelReturn ret;
+
+    controller_state = SEND_COMMAND;
+
+    while (controller_state != FINISH && controller_state != ERROR_STATE) {
+        switch (controller_state) {
+            case SEND_COMMAND: // Send start image transfer command
+            {
+                ret = send_command(IRIS_TRANSFER_IMAGE);
+                if (ret == IRIS_ACK) {
+                    controller_state = GET_DATA;
+                } else {
+                    controller_state = FINISH;
+                }
+                break;
+            }
+            case GET_DATA: // Get image data in chunks/blocks
+            {
+                uint16_t * image_data_buffer = (uint16_t*) pvPortMalloc(IMAGE_TRANSFER_SIZE * sizeof(uint16_t));
```

Comment:
Static declaration would be better here, since the size is known at compile time